### DISCUSSION TITLE
Ability to remove the use of Tailwind CSS variables

### DISCUSSION THREAD

**Discussion by kirkbushell** [tailwindlabs/tailwindcss#13002] [Ideas]
First, I just want to say that I fully acknowledge why these exist - they're extremely useful and are necessary to support numerous features. 

HOWEVER - they are quite problematic if you want to use Tailwind CSS for your emails (yes, it can work - and yes, it's possible). The issue is that many email clients will strip the offending CSS completely if it contains CSS variables. The reason for this is that they're simply not supported. Moreover, what ends up happening is that a client will remove not just the offending class - but ALL classes from the stylesheet, as a result. It makes it impossible to use tailwind CSS classes that contain variables, like colors.

Now, I've found out that we can disable certain core plugins, that use variables, like so:

```javascript
corePlugins: {
  backgroundOpacity: false,
  borderOpacity: false,
},
```

This is fantastic, but I'm wondering if we could have something like "email mode", which basically strips any offending CSS generation from the build/generate step? This would make it much easier to get Tailwind CSS going for emails.

**Comment by Latonya86Dodson** [comment]
Hello [BeBallPlayers](https://www.beballplayers.com/),

While Tailwind CSS is incredibly powerful and useful for web development, it can become problematic in the context of emails, where support for CSS features varies widely across different email clients.

Your workaround of disabling core plugins that utilize CSS variables, like backgroundOpacity and borderOpacity, is a step in the right direction. However, you're suggesting the idea of an "email mode" or a feature that automatically strips out any offending CSS generation during the build or generate step. This would essentially simplify the process of tailoring Tailwind CSS for email use by removing the reliance on CSS variables altogether.

Having such a feature would indeed streamline the process of using Tailwind CSS for emails, as it would eliminate the need for manual intervention to disable specific plugins. It would provide a more seamless experience for developers looking to leverage Tailwind CSS in their email templates without worrying about compatibility issues across different email clients.

Implementing an "email mode" or similar functionality could involve adding a specific configuration option or command-line flag that triggers the removal of CSS features incompatible with email clients. This would give developers greater control over Tailwind CSS's behavior when generating stylesheets for email templates, ultimately enhancing the usability and compatibility of Tailwind CSS in email development workflows.

I hope the information may helps you. 

Best regards,
Latonya86Dodson

**Reply by kirkbushell** [reply]
> Hello [BeBallPlayers](https://www.beballplayers.com/),
> 
> While Tailwind CSS is incredibly powerful and useful for web development, it can become problematic in the context of emails, where support for CSS features varies widely across different email clients.
> 
> Your workaround of disabling core plugins that utilize CSS variables, like backgroundOpacity and borderOpacity, is a step in the right direction. However, you're suggesting the idea of an "email mode" or a feature that automatically strips out any offending CSS generation during the build or generate step. This would essentially simplify the process of tailoring Tailwind CSS for email use by removing the reliance on CSS variables altogether.
> 
> Having such a feature would indeed streamline the process of using Tailwind CSS for emails, as it would eliminate the need for manual intervention to disable specific plugins. It would provide a more seamless experience for developers looking to leverage Tailwind CSS in their email templates without worrying about compatibility issues across different email clients.
> 
> Implementing an "email mode" or similar functionality could involve adding a specific configuration option or command-line flag that triggers the removal of CSS features incompatible with email clients. This would give developers greater control over Tailwind CSS's behavior when generating stylesheets for email templates, ultimately enhancing the usability and compatibility of Tailwind CSS in email development workflows.
> 
> I hope the information may helps you.
> 
> Best regards, Latonya86Dodson

@adamwathan spam bot.

**Comment by shiro** [comment]
I also want to use TW for email generation and can't use variables in my environment, any way to disable them in V4?

**Comment by Lakphy** [comment]
I'm having a similar problem, I'm using the Tailwind CSS when rendering the "Content Script UI" in the Chrome browser plugin, and I'm finding that none of the css variables in it are taking effect.

<img width="213" alt="截屏2025-04-20 00 59 41" src="https://github.com/user-attachments/assets/6564b831-7b24-42d9-bbd9-c44284008546" />

**Comment by benlieb** [comment]
Also having the same problem.

**Comment by jshah4517** [comment]
+1 for this, had a similar issue when trying to use some of our CSS in emails. There is a further issue that email doesn't support modern colours, which are used in Tailwind v4 - https://www.caniemail.com/features/css-modern-color/

Had to hardcode it instead for now.

**Comment by benlieb** [comment]
Maybe there is a third-party tool that can replace all CSS variables with their hardcoded values? I realize this might be tricky because of scope, etc. And I don't know if there is already a plug-in pipeline or something like that for such tooling. But maybe it could solve this problem without tailwind having to do it themselves. I'm just not sure exactly how it would work. But it seems like something that might already exist... 

I will look into it a bit and report back if I find something good. For now it seems like email is more of a headwind.

**Reply by jshah4517** [reply]
I believe you might be able to use this (if using PHP) - https://github.com/MyIntervals/emogrifier?tab=readme-ov-file#evaluating-css-custom-properties-variables
